modules/extensions-mule-support/src/main/java/org/mule/runtime/module/extension/mule/internal/loader/parser/MuleSdkExtensionExtensionModelParserFactory.java

```diff
@@ -0,0 +1,135 @@
+
+  private ArtifactAst cachedArtifactAst;
+
+  /**
+   * @param context The loading context.
+   * @return whether this factory is capable of creating a parser for the given {@link ExtensionLoadingContext}.
+   */
+  public static boolean handles(ExtensionLoadingContext context) {
+    // It needs a context containing a resource file.
+    if (context.getParameter(MULE_SDK_RESOURCE_PROPERTY_NAME).isPresent()) {
+      return true;
+    }
+
+    // ...or an already parsed AST that has an extension definition.
+    return context.<ArtifactAst>getParameter(MULE_SDK_ARTIFACT_AST_PROPERTY_NAME)
+        .map(ast -> ast.getArtifactType().equals(MULE_EXTENSION))
+        .orElse(true);
+  }
+
+  @Override
+  protected MuleSdkExtensionModelMetadataParser createMetadataParser(ExtensionLoadingContext context) {
+    // The existence of the name here means that the AST parser is trying to load the artifact-local ExtensionModel in order
+    // to complete the AST.
+    Optional<String> extensionName = context.getParameter(MULE_SDK_EXTENSION_NAME_PROPERTY_NAME);
+
+    if (extensionName.isPresent()) {
+      // In this case, we need to treat is as artifact-local, even when the artifact is an extension.
+      return new MuleSdkArtifactLocalExtensionModelMetadataParser(extensionName.get());
+    } else {
+      return new MuleSdkExtensionExtensionModelMetadataParser(getArtifactAst(context));
+    }
+  }
+
+  @Override
+  protected Supplier<Stream<ComponentAst>> createTopLevelComponentsSupplier(ExtensionLoadingContext context) {
+    // At this point we can assume there is only one top level component which is the extension:extension component
+    // We don't need to check for this because it should be guaranteed by previous validations
+    ComponentAst rootComponent = getArtifactAst(context).topLevelComponents().get(0);
+    return rootComponent::directChildrenStream;
+  }
+
+  private AstXmlParser createAstParser(Set<ExtensionModel> dependencies, boolean disableValidations) {
+    Builder astBuilder = builder()
+        .withArtifactType(MULE_EXTENSION)
+        .withExtensionModels(dependencies);
+
+    if (disableValidations) {
+      astBuilder.withSchemaValidationsDisabled();
+    }
+
+    return astBuilder.build();
+  }
+
+  private ArtifactAst parseAst(ExtensionLoadingContext context) throws ConfigurationException {
+    Set<ExtensionModel> dependencies = context.getDslResolvingContext().getExtensions();
+
+    String[] resources = {getMandatoryParameter(context, MULE_SDK_RESOURCE_PROPERTY_NAME)};
+    ArtifactAst artifactAst = parseArtifact(resources,
+                                            this::createAstParser,
+                                            dependencies,
+                                            false,
+                                            context.getExtensionClassLoader(),
+                                            "this", // should be irrelevant as long as it doesn't collide
```

Comment:
how wouldn't it collide if the value is hardcoded?

Reply:
It would be weird because there should be an extension in context named "this". However, I just changed it so that it uses a name which includes also the resource file name (which would be more helpful also for troubleshooting if there is any error during parsing).

In any case, keep in mind that this is only the name of the temporary Extension that gets build under the namespace "this" so that the full AST can get generated before doing the final loading of the real ExtensionModel. I explained a little bit more about this in the comments now.